src/compiler/crystal/codegen/call.cr

```diff
@@ -217,13 +219,26 @@ class Crystal::CodeGenVisitor
       case abi_arg_type.kind
       when LLVM::ABI::ArgKind::Direct
         call_arg = codegen_direct_abi_call(call_arg, abi_arg_type) unless arg.type.nil_type?
-        call_args << call_arg
       when LLVM::ABI::ArgKind::Indirect
         # Pass argument as is (will be passed byval)
-        call_args << call_arg
       when LLVM::ABI::ArgKind::Ignore
         # Ignore
+        next
+      end
+
+      # If we are passing variadic arguments there are some special rules
+      if i >= target_def_args_size
+        arg_type = arg.type.remove_indirection
+        if arg_type.is_a?(FloatType) && arg_type.kind == :f32
+          # Floats must be passed as doubles (there are no float varargs)
+          call_arg = extend_float @program.float64, call_arg
+        elsif arg_type.is_a?(IntegerType) && arg_type.kind.in?(:i8, :u8, :i16, :u16)
+          # Integer with a size less that `int` must be converted to `int`
+          call_arg = extend_int arg_type, @program.int32, call_arg
```

Comment:
So I guess technically this should check what the target platform C `int` type is, but in practice all platforms we support so far have `int = int32_t`. Just something to keep in mind in case somebody wants to target something weird or old.

Reply:
Yeah, I thought of this and reached the same conclusion. Also, I don't know how to know what's the target platform size of `int`.

Reply:
`sizeof(LibC::Int)` ?

Reply:
That's the size of the host machine. It won't work when cross-compling.